### PR TITLE
Refactor Docker builds to use Depot for zero emulation builds + persistent cache

### PR DESCRIPTION
**Pull request by kylegalbraith**
## Summary
This moves the Docker image builds to [Depot](https://depot.dev), where they can be done on native Intel and Arm CPUs with persistent layer cache across builds.

## Background
I sent our [benchmark-astria results](https://github.com/depot/benchmark-astria/actions/runs/8647400279/job/23708761387) over to @joroshiba that shows how Depot can build all these images in ~46 seconds for a fully cached build. He asked if we could open a PR to help with the cutover.

## Changes
- `reusable-docker-build.yml` is updated to use [`depot/build-push-action`](https://github.com/depot/build-push-action), which takes all the same parameters with the addition of a `project` property to route the image build to a specific project in Depot.
- `reusable-docker-build.yml` contains a new permissions block so that you can authenticate to Depot from a GHA job via an OIDC access token instead of having static tokens in CI (see [trust relationship docs](https://depot.dev/docs/cli/authentication#oidc-trust-relationships))
- Drops `cache-to` and `cache-from` as Depot projects come with persistent caching to real SSDs that are automatically orchestrated across builds.
- `docker-build.yml` is updated to pass in the Depot project ID for each image to build.
- Run the `reusable-docker-build.yml` on a default GHA runner as you no longer need the BuildJet 8 CPU runner because the image build is happening on a Depot image builder with 16 CPUs & 32 GB of RAM & a 50 GB NVMe SSD for each architecture.

## Testing
Tested the basic components of this in the benchmark.

### CODE REVIEW

File: .github/workflows/docker-build.yml [hunk 1]
```diff
@@ -43,8 +42,12 @@ jobs:
   composer:
     needs: run_checker
     if: needs.run_checker.outputs.run_docker == 'true' || (github.event_name == 'workflow_dispatch' && github.event.inputs.target == 'composer')
-    uses: './.github/workflows/reusable-docker-build.yml'
+    uses: "./.github/workflows/reusable-docker-build.yml"
+    permissions:
+      contents: read
+      id-token: write
     with:
+      depot-project-id: TODO
```
Comment (joroshiba) [inline .github/workflows/docker-build.yml:50]:
```suggestion
      depot-project-id: mhgvgvsjnx
```

File: .github/workflows/docker-build.yml [hunk 2]
```diff
@@ -53,8 +56,12 @@ jobs:
   conductor:
     needs: run_checker
     if: needs.run_checker.outputs.run_docker == 'true' || (github.event_name == 'workflow_dispatch' && github.event.inputs.target == 'conductor')
-    uses: './.github/workflows/reusable-docker-build.yml'
+    uses: "./.github/workflows/reusable-docker-build.yml"
+    permissions:
+      contents: read
+      id-token: write
     with:
+      depot-project-id: TODO
```
Comment (joroshiba) [inline .github/workflows/docker-build.yml:64]:
```suggestion
      depot-project-id: zrh9t1d84s
```

File: .github/workflows/docker-build.yml [hunk 3]
```diff
@@ -63,8 +70,12 @@ jobs:
   sequencer:
     needs: run_checker
     if: needs.run_checker.outputs.run_docker == 'true' || (github.event_name == 'workflow_dispatch' && github.event.inputs.target == 'sequencer')
-    uses: './.github/workflows/reusable-docker-build.yml'
+    uses: "./.github/workflows/reusable-docker-build.yml"
+    permissions:
+      contents: read
+      id-token: write
     with:
+      depot-project-id: TODO
```
Comment (joroshiba) [inline .github/workflows/docker-build.yml:78]:
```suggestion
      depot-project-id: brzhxfbv9b
```

File: .github/workflows/docker-build.yml [hunk 4]
```diff
@@ -73,13 +84,17 @@ jobs:
   sequencer-relayer:
     needs: run_checker
     if: needs.run_checker.outputs.run_docker == 'true' || (github.event_name == 'workflow_dispatch' && github.event.inputs.target == 'sequencer-relayer')
-    uses: './.github/workflows/reusable-docker-build.yml'
+    uses: "./.github/workflows/reusable-docker-build.yml"
+    permissions:
+      contents: read
+      id-token: write
     with:
+      depot-project-id: TODO
```
Comment (joroshiba) [inline .github/workflows/docker-build.yml:92]:
```suggestion
      depot-project-id: 86q4kz4wfs
```

File: .github/workflows/reusable-docker-build.yml [hunk 5]
```diff
@@ -70,10 +76,8 @@ jobs:
           file: containerfiles/Dockerfile
           build-args: |
             TARGETBINARY=${{ inputs.target-binary }}
-          platforms: 'linux/amd64,linux/arm64'
-          push: true
+          platforms: "linux/amd64,linux/arm64"
+          push: ${{ env.LOGIN_DOCKER || 'false' }}
```
Comment (kylegalbraith) [inline .github/workflows/reusable-docker-build.yml:80]: Do fork PRs get pushed to the registries today? I wasn't sure about that one. If not, I think we could update this to check if the job is from a fork and not push.

```yaml
push: ${{ github.event.pull_request.head.repo.full_name == 'org/repo' }}
```

Comment (joroshiba) [reply .github/workflows/reusable-docker-build.yml]: They don't, I updated to validating that it's owned by astriaorg

**Review by joroshiba — APPROVED** [review]

**Pull request merged**